### PR TITLE
fix(rust): Reject non-integral offset and length in AExpr::Slice

### LINKED ISSUE #16855: [QST]: should `pl.Expr.slice` raise if the numeric arguments are not integral
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
df = pl.DataFrame({"a": [1, 2, 3, 4]})

q = df.select(pl.col.a.slice(1.7, 2))
```

### Log output

_No response_

### Issue description

Expression (series) slicing complains if the offset and length are not literals, or not numeric types. But non-integral numeric values are happily accepted.

### Expected behavior

I was expecting this to raise, since slicing from 1.7 doesn't make much sense.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.31
Index type:           UInt32
Platform:             macOS-14.5-arm64-arm-64bit
Python:               3.11.7 (main, Dec  4 2023, 18:10:11) [Clang 15.0.0 (clang-1500.1.0.2.5)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           <not installed>
nest_asyncio:         <not installed>
numpy:                <not installed>
openpyxl:             <not installed>
pandas:               <not installed>
pyarrow:              <not installed>
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
torch:                <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by wence-**
- Fixes #16855

### CODE REVIEW

File: py-polars/tests/unit/dataframe/test_df.py [hunk 1]
```diff
@@ -1976,7 +1976,7 @@ def test_group_by_slice_expression_args() -> None:
 
     out = (
         df.group_by("groups", maintain_order=True)
-        .agg([pl.col("vals").slice(pl.len() * 0.1, (pl.len() // 5))])
+        .agg([pl.col("vals").slice((pl.len() * 0.1).cast(int), (pl.len() // 5))])
```
Comment (wence-) [inline py-polars/tests/unit/dataframe/test_df.py:1979]: This usage is now rejected, so we must explicitly cast.

File: py-polars/tests/unit/expr/test_exprs.py [hunk 2]
```diff
@@ -690,7 +690,7 @@ def test_tail() -> None:
     assert df.select(pl.col("a").tail(10)).to_dict(as_series=False) == {
         "a": [1, 2, 3, 4, 5]
     }
-    assert df.select(pl.col("a").tail(pl.len() / 2)).to_dict(as_series=False) == {
+    assert df.select(pl.col("a").tail(pl.len() // 2)).to_dict(as_series=False) == {
```
Comment (wence-) [inline py-polars/tests/unit/expr/test_exprs.py:693]: Floor division to maintain integral nature.

File: py-polars/polars/expr/expr.py [hunk 3]
```diff
@@ -4851,7 +4851,11 @@ def tail(self, n: int | Expr = 10) -> Self:
         │ 7   │
         └─────┘
         """
-        offset = -self._from_pyexpr(parse_as_expression(n))
+        # This cast enables tail with expressions that return unsigned integers,
+        # for which negate otherwise raises InvalidOperationError.
+        offset = -self._from_pyexpr(
+            parse_as_expression(n).cast(Int64, strict=False, wrap_numerical=True)
+        )
```
Comment (wence-) [inline py-polars/polars/expr/expr.py:4858]: This is arguably a "feature" improvement. Previously one could not ask for the `tail` for an expression with an expression value of unsigned integer type. That is, this enables the following:

```
df.select(pl.col("a").tail(pl.lit(1, pl.UInt32()))
```

The casting is similar to the rust side implementation of tail:
```rust
    /// Get the last `n` elements of the Expr result.
    pub fn tail(self, length: Option<usize>) -> Self {
        let len = length.unwrap_or(10);
        self.slice(lit(-(len as i64)), lit(len as u64))
    }
```

**Review by wence- — COMMENTED** [review]
Signposts

File: crates/polars-plan/src/logical_plan/conversion/type_coercion/mod.rs [hunk 4]
```diff
@@ -440,6 +440,20 @@ impl OptimizationRule for TypeCoercionRule {
                     options,
                 })
             },
+            AExpr::Slice { offset, length, .. } => {
+                let input_schema = get_schema(lp_arena, lp_node);
+                let (_, offset_dtype) =
+                    unpack!(get_aexpr_and_type(expr_arena, offset, &input_schema));
+                if !offset_dtype.is_integer() {
+                    polars_bail!(InvalidOperation: "offset must be integral for slice, not {}", offset_dtype)
```
Comment (ritchie46) [inline crates/polars-plan/src/logical_plan/conversion/type_coercion/mod.rs:448]: Can we use `polars_ensure`?

File: crates/polars-plan/src/logical_plan/conversion/type_coercion/mod.rs [hunk 5]
```diff
@@ -440,6 +440,20 @@ impl OptimizationRule for TypeCoercionRule {
...
+                }
+                let (_, length_dtype) =
+                    unpack!(get_aexpr_and_type(expr_arena, length, &input_schema));
+                if !length_dtype.is_integer() {
```
Comment (ritchie46) [inline crates/polars-plan/src/logical_plan/conversion/type_coercion/mod.rs:452]: Can we use polars_ensure?

Comment (wence-) [reply crates/polars-plan/src/logical_plan/conversion/type_coercion/mod.rs]: Done.

**Comment by ritchie46** [comment]
Ah.. Sorry for the inconvenience, but can you rebase? Seems to be some conflicts.

**Comment by wence-** [comment]
> Ah.. Sorry for the inconvenience, but can you rebase? Seems to be some conflicts.

Done.

**Pull request merged**